app/models/user_batch.rb

```diff
@@ -35,79 +36,247 @@ def create_users(mission)
+             reset_password_method: "print",
+             admin: false,
+             assignments: [Assignment.new(mission: mission, role: User::ROLES.first)],
+             skip_validations_with_queries: true,
+             import_num: @import_num += 1))
+  end
 
-        # skip blank rows
-        next if row.all?(&:blank?)
-
-        # turn the row into an attribute hash
-        attributes = Hash[*row.map do |cell|
-          field = fields[cell.coordinate.column - 1]
-          [field, cell.value.presence]
-        end.flatten]
-
-        # attempt to create the user with the parsed params
-        user = User.create(attributes.merge(
-          reset_password_method: "print",
-          assignments: [Assignment.new(mission_id: mission.id, role: User::ROLES.first)]))
-
-        # if the user has errors, add them to the batch's errors
-        if !user.valid?
-          user.errors.keys.each do |attribute|
-            user.errors.full_messages_for(attribute).each do |error|
-              row_error = I18n.t('import.row_error', row: row_number, error: error)
-              errors.add("users[#{row_index}].#{attribute}", row_error)
-            end
+  def add_validation_error_messages(user, row_number)
+    # if the user has errors, add them to the batch's errors
+    unless user.errors.empty?
+      user.errors.keys.each do |attribute|
+        if attribute != :persistence_token
+          user.errors.full_messages_for(attribute).each do |error|
+            add_error(error, attribute, row_number)
           end
-          @validation_error = true
         end
+      end
+      @validation_error = true unless error_is_on_persistence_token? user
+    end
+  end
 
-        users << user
+  def error_is_on_persistence_token?(user)
+    (user.errors.keys.length == 1) && (user.errors.keys.include? :persistence_token)
+  end
 
-      end # iteration over rows
+  def add_error(error, attribute, row_number)
+    row_error = I18n.t('import.row_error', row: row_number, error: error)
+    errors.add("users[#{row_number}].#{attribute}", row_error)
+  end
 
-      # now if there was a validation error with any user, rollback the transaction
-      raise ActiveRecord::Rollback if @validation_error
+  def add_too_many_errors(row_number)
+    # we pass the previous row number to the error message formatting
+    # since the error count cutoff was surpassed by the previous rows,
+    # not this row
+    errors.add(:users, :too_many_errors, row: row_number - 1) if errors_reached_limit
+  end
 
-    end # transaction
+  def errors_reached_limit
+    errors.count >= IMPORT_ERROR_CUTOFF
+  end
 
-    return succeeded?
+  def actual_row_number(row_start, index)
+    index + row_start + 1
+  end
+
+  def check_uniqueness_on_objects(objects, field)
+    key = correct_field_key(field)
+    # Add errors on fields that aren't unique
+    @fields_hash_table[key].each do |k,v|
+      v.each { |i| objects.at(i).errors.add(field, :taken) } if v.length != 1
```

Comment:
No space after `.each` please.

Also, why not just add the objects themselves to the hash table instead of their array index? Then you save the kind of awkward `.at(i)` thing, which was a bit hard to decipher.